Support webtunnel, obfs and snowflake bridges

akrigator opened the pull request:
# Proposed Changes

One more option to resolve tor network connectivity issue.

### Dockerfile
* install Go and dependent packages
* build obfs4proxy and snowflake client by Go

### Scripts
* Push bridge configuration from addon's configuration to `/etc/tor/torrc` before the first tor execution

## Related Issues


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced new version management for obfs4, snowflake, and webtunnel in the Dockerfile.
  - Added support for HTTP tunneling and bridge configurations in the Tor add-on.

- **Enhancements**
  - Updated package installations in the Dockerfile for improved setup.
  - Extended health check start period from 5 minutes to 15 minutes for better initial setup.

- **Configuration Updates**
  - Added regex managers for managing versions of obfs4, snowflake, and webtunnel in the Renovate configuration.
  - Introduced new configuration options for HTTP proxy and bridges in the Tor add-on documentation and configuration files.
  - Expanded available ports and options in the Tor add-on configuration.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
Update Dockerfile

install obfs proxy

Commit:
Update config.yaml

Default bridges

Commit:
Update run

Commit:
Add obfs4 bridges support

akrigator commented:
Hello, @frenck could you help to proceed with pr

Commit:
Merge remote-tracking branch 'upstream/main' into obfs

Commit:
Merge remote-tracking branch 'upstream/main' into obfs

Commit:
Update transaltion

Commit:
Markdown experiment

Commit:
Move bridge link to doc from translation

Commit:
Move bridge link to doc from translation

Commit:
Snowflake  bridge support

Commit:
Move defualt config to docs, update translation

Commit:
Merge remote-tracking branch 'upstream/main' into obfs

Commit:
Passing snowflake_* options throw bridge's option

Commit:
Merge remote-tracking branch 'upstream/main' into obfs

Commit:
Fix dependcy version

Snowflake 2.7.0 is dependent on go 1.21+.
Let's use 2.6.1 since only 1.20.10-r0 is available in ghcr.io/hassio-addons/base:14.3.0

Commit:
Update docs and fix snowflake port binding privelage

Commit:
Add webtunnel bidge support

Commit:
Increase watchdog interval

Bootstrapping with bridge takes upto 1 hour
https://docs.docker.com/engine/reference/builder/#healthcheck

Commit:
Merge remote-tracking branch 'upstream/main' into obfs

Commit:
Merge remote-tracking branch 'upstream/main' into obfs

frenck commented on tor/Dockerfile:
```diff
@@ -13,10 +13,39 @@ RUN \
+        libcap=2.69-r0
+
+# Install obfs4proxy transport client
+RUN git clone https://github.com/Yawning/obfs4.git /go/obfs4/
```
- Please clone a specific version
- Add the version as a variable and configure renovate to update it.

akrigator replied:
Fixed

frenck commented on tor/Dockerfile:
```diff
@@ -13,10 +13,39 @@ RUN \
+RUN go build -o /usr/local/bin/obfs4proxy ./obfs4proxy
+
+# Install snowflake transport client
+RUN git clone --branch v2.6.1 https://gitlab.torproject.org/tpo/anti-censorship/pluggable-transports/snowflake.git /go/snowflake/
+WORKDIR /go/snowflake/client/
+RUN go get
+RUN go build -o /usr/local/bin/snowflake
```
Similar to above

We could just move this into the above sequence, we don't need extra Docker layers in this case.

frenck commented on tor/Dockerfile:
```diff
@@ -13,10 +13,39 @@ RUN \
+RUN go build -o /usr/local/bin/snowflake
+
+# Install webtunnel client
+RUN git clone https://gitlab.torproject.org/tpo/anti-censorship/pluggable-transports/webtunnel.git /go/webtunnel/
+WORKDIR /go/webtunnel/main/client/
+RUN go build -ldflags="-s -w" -o /usr/local/bin/webtunnel
```
Same as above

frenck commented on tor/Dockerfile:
```diff
@@ -13,10 +13,39 @@ RUN \
+
+# Give tranposrts clients the capability to bind port 80. This line isn't necessary if
+# you use a high (unprivileged) port.
+RUN setcap \
```
This can just be part of the above sequence without creating an extra Docker layer

frenck commented on tor/Dockerfile:
```diff
@@ -13,10 +13,39 @@ RUN \
 
 HEALTHCHECK \
-    --start-period=5m \
+    --start-period=60m \
```
Why 60! minutes?

akrigator replied:
Connection to bridge takes up to 60 minutes. Other way Docker restarts  container every  5 minutes during bridge bootstrapping, and it will never connect to tor. 
I guess it can be user configurable or switch between two predefined values depending on the bridge configuration

akrigator replied:
So after year of usage and monitoring this period i could say that in most cases 15 minutes will be enough, but it really depends on transport plugins implementation.

Commit:
Fix up @frenck comments:

- setup renovate to manage tor trasports plugins versioning
- install transport plugins in single docker layer

Commit:
Merge remote-tracking branch 'origin/main' into obfs

Commit:
Fix up @frenck comments:

- setcap in single docker layer

Commit:
Merge remote-tracking branch 'upstream/main' into obfs

Commit:
Merge branch 'main' into obfs

Commit:
Merge remote-tracking branch 'upstream/main' into obfs

Commit:
Update transports and their dependencies

Commit:
Merge remote-tracking branch 'upstream/main' into obfs

Commit:
Docker layers

Commit:
Docker layers

Commit:
YAMLLint fixup

akrigator committed:
Prettified Code!

Commit:
Docker layers

Commit:
Merge remote-tracking branch 'origin/refactoring' into refactoring

Commit:
Docker layers

Commit:
Misspelling

Commit:
Remove bridge sources from image

Commit:
coderabbitai suggestions

Commit:
Revert suggestion

Commit:
Merge remote-tracking branch 'upstream/main' into obfs

Commit:
Update dependecies

Commit:
Update dependency

Commit:
Update dependincies

Commit:
Update dependency

Commit:
Merge remote-tracking branch 'upstream/main' into obfs

Commit:
Update dependency packages

Commit:
Update web tunnel version

akrigator committed:
Merge branch 'main' into obfs

frenck commented:
Ok let's get this in, it is hanging for too long (totally my fault), can you rebase?

Commit:
Merge change from the alexbelgium/hassio-addons tor version

Commit:
Merge remote-tracking branch 'upstream/main' into obfs

akrigator committed:
Prettified Code!

Commit:
Decrease healthcheck start period

Commit:
Fixup documentation

Commit:
Remove trailing spaces

akrigator commented:
Hello @frenck , thank you for find time for this tiny pr. I've updated it with changes from recent change in https://github.com/alexbelgium/hassio-addons/pull/1772 and your main branch

frenck commented on tor/DOCS.md:
```diff
@@ -85,6 +87,12 @@ applications on your network to use the Tor network.
 
+### Option: `http_tunnel`
+
+Setting this option to `true` opens port `9080` to listen for connections from
+HTTP-speaking applications. Enabling this feature allows you to use other
+applications on your network to use the Tor network via http proxy.
```
```suggestion
Setting this option to `true` opens port `9080` to listen for connections from
HTTP-speaking applications. Enabling this feature allows you to use other
applications on your network to access the Tor network via the HTTP proxy.
```

frenck commented on tor/Dockerfile:
```diff
@@ -12,11 +12,46 @@ COPY rootfs /
+        libcap=2.71-r0
+
+# Download pluggable-transports sources
+WORKDIR /go
+ARG OBFS_VERSION=obfs4proxy-0.0.14
+ARG SNOWFLAKE_VERSION=v2.10.1
+ARG WEBTUNNEL_VERSION=v0.0.2
+RUN git clone -b ${OBFS_VERSION} --single-branch --depth 1 https://github.com/Yawning/obfs4.git \
+    && git clone -b ${SNOWFLAKE_VERSION} --single-branch --depth 1 https://gitlab.torproject.org/tpo/anti-censorship/pluggable-transports/snowflake.git \
+    && git clone -b ${WEBTUNNEL_VERSION} --single-branch --depth 1 https://gitlab.torproject.org/tpo/anti-censorship/pluggable-transports/webtunnel.git
+
+# Build obfs4proxy
+WORKDIR /go/obfs4
+RUN go build -o /usr/local/bin/obfs4proxy ./obfs4proxy
+
+# Build snowflake
+WORKDIR /go/snowflake/client
+RUN go get \
+    && go build -ldflags '-extldflags "-static" -w -s' -o /usr/local/bin/snowflake
+
+# Build webtunnel
+WORKDIR /go/webtunnel/main/client
+RUN go get \
+    && go build -ldflags="-s -w" -o /usr/local/bin/webtunnel
+
+# Clean up after build
+WORKDIR /
+RUN rm -rf /go
+
+# Give transports clients the capability to bind privileged port.
+RUN setcap 'cap_net_bind_service=+ep' /usr/local/bin/obfs4proxy \
+    && setcap 'cap_net_bind_service=+ep' /usr/local/bin/snowflake \
+    && setcap 'cap_net_bind_service=+ep' /usr/local/bin/webtunnel
```
Let's please avoid all the layer creation and chain this into a single layer.

akrigator replied:
Could we ignore this https://github.com/akrigator/addon-tor/actions/runs/13379472629/job/37365356801 ?

frenck commented on tor/Dockerfile:
```diff
@@ -12,11 +12,46 @@ COPY rootfs /
-        openssl=3.3.3-r0 \
-        tor=0.4.8.14-r1
+        tor=0.4.8.14-r1 \
+        go=1.23.6-r0 \
```
Some of the added dependencies are build-only dependencies. They need to be installed temporarily and cleaned up after the build.

Otherwise they will end up being death weight in the final resulting image.

akrigator replied:
Actually `ca-certificates` and `libcap` is required only for server-side of transport. Here we use only their clients, so that why those packages are redundant at all. Good catch)
I remember how have tried to build clients by server's guide that's why it was missed here.

frenck replied:
Let's handle build dependencies like this:

https://github.com/hassio-addons/addon-node-red/blob/b8090b301fed7cb1b3d94051cd377e1e6282a20a/node-red/Dockerfile#L17-L22

frenck commented on tor/rootfs/etc/s6-overlay/s6-rc.d/init-tor/run:
```diff
@@ -89,6 +89,11 @@ else
 fi
 
+# Configure Http tunnel port
+if bashio::config.true 'http_tunnel'; then
```
Instead of introducing another option, we can also check if the port is actually set in the add-on configuration (and default it to `null`).

frenck replied:
Ignore this one, you are mimicking the existing socks 👍

frenck commented on tor/translations/en.yaml:
```diff
@@ -30,5 +30,16 @@ configuration:
+  bridges:
+    name: Bridges
+    description: >-
+      Configures bridges to establish connection with Tor Network. Check the
+      add-on documentation for the exact format to enter here.
```
```suggestion
      Configures bridges to establish a connection with the Tor network. Check the
      add-on documentation for the exact format to enter here.
```

akrigator committed:
Update tor/DOCS.md

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

akrigator committed:
Update tor/translations/en.yaml

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

Commit:
Refactoring by feedback

Commit:
9 layers

Commit:
3 layers

frenck commented on tor/Dockerfile:
```diff
@@ -8,15 +8,31 @@ SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+        tor=0.4.8.14-r1 \
+        go=1.23.6-r0 \
+        git=2.47.2-r0 \
+    && git clone -b ${OBFS_VERSION} --single-branch --depth 1 \
```
```suggestion
    && git clone -b "${OBFS_VERSION}" --single-branch --depth 1 \
```

frenck commented on tor/Dockerfile:
```diff
@@ -8,15 +8,31 @@ SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+    && git clone -b ${OBFS_VERSION} --single-branch --depth 1 \
+        https://github.com/Yawning/obfs4.git /go/obfs4 \
+    && go build -C /go/obfs4 -ldflags '-extldflags "-static" -s -w' -o /usr/local/bin/obfs4proxy ./obfs4proxy \
+    && git clone -b ${SNOWFLAKE_VERSION} --single-branch --depth 1 \
```
```suggestion
    && git clone -b "${SNOWFLAKE_VERSION}" --single-branch --depth 1 \
```

frenck commented on tor/Dockerfile:
```diff
@@ -8,15 +8,31 @@ SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+    && git clone -b ${SNOWFLAKE_VERSION} --single-branch --depth 1 \
+        https://gitlab.torproject.org/tpo/anti-censorship/pluggable-transports/snowflake.git /go/snowflake \
+    && go build -C /go/snowflake/client -ldflags '-extldflags "-static" -s -w' -o /usr/local/bin/snowflake \
+    && git clone -b ${WEBTUNNEL_VERSION} --single-branch --depth 1 \
```
```suggestion
    && git clone -b "${WEBTUNNEL_VERSION}" --single-branch --depth 1 \
```

frenck commented on tor/Dockerfile:
```diff
@@ -8,15 +8,31 @@ SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+    && git clone -b ${WEBTUNNEL_VERSION} --single-branch --depth 1 \
+        https://gitlab.torproject.org/tpo/anti-censorship/pluggable-transports/webtunnel.git /go/webtunnel \
+    && go build -C /go/webtunnel/main/client -ldflags='-extldflags "-static" -s -w' -o /usr/local/bin/webtunnel \
+    && rm -rf /go \
```
Let's put tmp stuff into the `/tmp` folder.

akrigator committed:
Apply suggestions from code review

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

Commit:
Apply suggestions from code review

Commit:
Apply suggestions from code review

frenck reviewed (approved):
Thanks, @akrigator 👍 

../Frenck
